### ISSUE TITLE
[API] Invite endpoint

### ISSUE THREAD

**Issue by Commifreak** [rdmorganiser/rdmo#540]
### Rationale / Begründung

API endpoint "Invite" to create/manage user invitations to a project

### Affected 

Users

### Minimal functionality

general CRUD operations

### Nice-to-have functionality

-

**Comment by jochenklar** [comment]
Hi @Commifreak , what you can already do is add users with the `/api/v1/projects/projects/<project_id>/memberships/` endpoint. What we would need to do is to allow the email invite, via the API.

**Comment by jochenklar** [comment]
Until we implement this, which could take some time, you could make your API user as `site_manager`, then it should be able to post to the invite form for every project. You probably need to fetch the form first for the csrf token. A bit hacky, I know.

**Comment by Commifreak** [comment]
Okay. I build a simple mechanism for now. If anyone would need that:

```php
/**
 * Create email invitation for an unknown user
 * @param $project_id integer Project-ID
 * @param $userMail string User Mail
 *
 * @return bool|void
 */
public static function sendInvitation($project_id, string $userMail) {
	$cookies = tempnam(sys_get_temp_dir(), 'rdmo_cookies');
	$baseUrl = \Yii::$app->params['rdmo']['endpoint'];

	/**
	 * Get Login CSRF Token
	 */
	$ch      = curl_init( $baseUrl . "/account/login/" );
	curl_setopt( $ch, CURLOPT_RETURNTRANSFER, true );
	curl_setopt( $ch, CURLOPT_COOKIEJAR, $cookies ); # store
	$result = curl_exec( $ch );
	curl_close( $ch );

	if ( ! $result ) {
		return false;
	}

	/**
	 * Parse HTML Body for token
	 */
	$csrf = self::getCSRFToken( $result );

	if ( ! $csrf ) {
		return false;
	}


	/**
	 * Do the actual login
	 */
	$ch = curl_init( $baseUrl . "/account/login/" );
	curl_setopt( $ch, CURLOPT_RETURNTRANSFER, true );
	curl_setopt( $ch, CURLOPT_COOKIEJAR, $cookies ); # store
	curl_setopt( $ch, CURLOPT_COOKIEFILE, $cookies ); # read & send
	curl_setopt( $ch, CURLOPT_HTTPHEADER, [ # Important for CSRF Validation
		'Referer: '.$baseUrl . "/account/login/"
	] );

	$data = [
		'csrfmiddlewaretoken' => $csrf,
		'username' => MasterSetting::get('rdmo_username'),
		'password' => MasterSetting::get('rdmo_password')
	];
	curl_setopt($ch, CURLOPT_POSTFIELDS, $data);

	$result = curl_exec($ch);
	curl_close($ch);

	if ( ! $result ) {
		return false;
	}


	/**
	 * Get Invitation form CSRF token
	 */
	$ch = curl_init( $baseUrl . "/projects/$project_id/memberships/create/" );
	curl_setopt( $ch, CURLOPT_RETURNTRANSFER, true );
	curl_setopt( $ch, CURLOPT_COOKIEJAR, $cookies );
	curl_setopt( $ch, CURLOPT_COOKIEFILE, $cookies );

	$result = curl_exec($ch);
	curl_close($ch);

	if ( ! $result ) {
		return false;
	}

	$csrf = self::getCSRFToken($result);

	if ( ! $csrf ) {
		return false;
	}


	/**
	 * Invite actually
	 */

	$ch = curl_init( $baseUrl . "/projects/$project_id/memberships/create/");
	curl_setopt( $ch, CURLOPT_RETURNTRANSFER, true );
	curl_setopt( $ch, CURLOPT_COOKIEJAR, $cookies );
	curl_setopt( $ch, CURLOPT_COOKIEFILE, $cookies );
	curl_setopt( $ch, CURLOPT_HTTPHEADER, [
		'Referer: '.$baseUrl . "/projects/$project_id/memberships/create/"
	] );

	$data = [
		'csrfmiddlewaretoken' => $csrf,
		'next' => '',
		'username_or_email' => $userMail,
		'role' => 'owner'
	];
	curl_setopt($ch, CURLOPT_POSTFIELDS, $data);

	$result = curl_exec($ch);
	$info = curl_getinfo($ch, CURLINFO_HTTP_CODE);
	curl_close($ch);

	unlink($cookies);

	if($info != 302) {
		return false;
	}

	return true;

}

/**
 * Parse CSRF Token from a HTML body
 * @param string $html The HTML Body
 *
 * @return false|string
 */
private static function getCSRFToken( string $html ) {
	$needle       = 'csrfmiddlewaretoken" value="';

	$csrfFieldPos = strpos( $html, $needle );
	if ( ! $csrfFieldPos ) {
		return false;
	}

	$csrfValueStop = strpos( $html, '"', $csrfFieldPos + strlen( $needle ) );
	if ( ! $csrfValueStop ) {
		return false;
	}

	$token = substr( $html, $csrfFieldPos + strlen( $needle ), $csrfValueStop-strlen( $needle ) -  $csrfFieldPos);

	if ( ! $token || strlen( $token ) != 64 ) {
		return false;
	}

	return $token;
}
```

**Comment by Commifreak** [comment]
EDIT: Ooops, found an error in the code. Just corrected.

**Comment by MyPyDavid** [comment]
the PR was merged in the 1.11 release, can this be closed?

**Comment by Commifreak** [comment]
I believe so
